### ISSUE TITLE
kill(ratvars) should give an error

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#2682]
Imported from SourceForge on 2024-07-06 17:16:42
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2018-05-22 19:44:46
Original: https://sourceforge.net/p/maxima/bugs/3434

---

kill(ratvars) should give an error, just as assigning it to an illegal value does.

An unbound ratvars causes problems....

Maxima 5.41.0a_dirty http://maxima.sourceforge.net
using Lisp SBCL 1.3.18
Distributed under the GNU Public License. See the file COPYING.
Dedicated to the memory of William Schelter.
The function bug_report() provides bug reporting information.
(%i1) display2d:false;leftjust:true$

(%o1) false
(%i3) ratvars;

(%o3) []
(%i4) rat(3);

(%o4) 3
(%i5) ratvars:23;

assignment: cannot assign 23 to ratvars                    <<<< Good check!
 -- an error. To debug this try: debugmode(true);
(%i6) rat(3);

(%o6) 3
(%i7) kill(ratvars);

remvalue: ratvars
           doesn't appear to be a known variable; just unbind it anyway.   <<<< NO!!!
(%o7) done
(%i8) rat(3);

Maxima encountered a Lisp error:

 The variable $RATVARS is unbound.

Automatically continuing.
To enable the Lisp debugger set *debugger-hook* to nil.
(%i9) quit();                                                              <<< Can't even QUIT if ratvars is killed

Maxima encountered a Lisp error:

 The variable $RATVARS is unbound.

Automatically continuing.
To enable the Lisp debugger set *debugger-hook* to nil.
(%i10)

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:16:43
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2021-01-12 03:50:51
Original: https://sourceforge.net/p/maxima/bugs/3434/#8f03

---

- **labels**:  --> kill, ratvars

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:16:47
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2021-01-12 03:50:52
Original: https://sourceforge.net/p/maxima/bugs/3434/#7cea

---

I wonder if KILL1-ATOM honors a list of unkillable symbols ... anyway I guess that's the first thing to look for.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:16:50
Created by **[peterpall](https://sourceforge.net/u/peterpall/)** on 2021-01-12 17:44:27
Original: https://sourceforge.net/p/maxima/bugs/3434/#7cea/fc91

---

That list would be long and - as it would have to be compiled manually - possibly incomplete: Ever tried to kill `inchar` or `outchar`?  My hope is that perhaps all  vital built-in symbols share a common property of some sort.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:16:54
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2021-01-12 18:16:09
Original: https://sourceforge.net/p/maxima/bugs/3434/#9693

---

The appropriate property is 'assign. Unfortunately, inchar/outchar don't have assign properties. They should, and they should be added to msetchk:

OLD:
	((member x '($genindex $optimprefix) :test #'eq) (if (not (symbolp y)) (mseterr x y)))
    
NEW:
	((member x '($inchar $outchar $genindex $optimprefix)) (if (not (symbolp y)) (mseterr x y)))

This is also an opportunity to get rid of the idiotic :test #'eq, which was mechanically added years ago by someone who didn't understand Lisp.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:16:57
Created by **[peterpall](https://sourceforge.net/u/peterpall/)** on 2021-01-12 18:29:46
Original: https://sourceforge.net/p/maxima/bugs/3434/#9693/a6f1

---

Another example of a thing one should not kill would be `logexpand`.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:17:01
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2021-01-12 19:00:04
Original: https://sourceforge.net/p/maxima/bugs/3434/#64a3

---

Looking at the source, here are some more cases with the legal values that should be checked by msetchk/assign:

$logexpand
(nil t $all $super)

$loadprint
(nil $loadfile)

$radexpand $rootsconmode $triginverses $scalarmatrixp $assumescalar
(nil t $all)

$savedef $showtime
(nil t $all) -- not documented

$tr_warn_fexpr $tr_warn_meval $tr_warn_mode $tr_warn_undeclared $tr_warn_undefined_variable
($all $compile $compfile $translate)

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:17:04
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2021-01-12 20:18:32
Original: https://sourceforge.net/p/maxima/bugs/3434/#ddd8

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:17:08
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2021-01-12 20:18:32
Original: https://sourceforge.net/p/maxima/bugs/3434/#0ba6

---

Fixed by commit 1d6b5c8, which adds this line to KILL1 in src/suprv1.lisp:
```{patch}
diff --git a/src/suprv1.lisp b/src/suprv1.lisp
index 7108f4a09..96ecc2d9a 100644
--- a/src/suprv1.lisp
+++ b/src/suprv1.lisp
@@ -321,6 +321,7 @@
              (mapc #'(lambda (y) (kill1 (caar y))) (cdr (symbol-value x))))
             ((eq x '$myoptions))
             ((eq x '$tellrats) (setq tellratlist nil))
+            ((eq x '$ratvars) (setq $ratvars '((mlist simp)) varlist nil))
             ((eq x '$ratweights) (setq *ratweights nil
                                        $ratweights '((mlist simp))))
             ((eq x '$features)
```

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:17:11
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2021-01-12 20:21:13
Original: https://sourceforge.net/p/maxima/bugs/3434/#1c2e

---

I see that this report has attracted some comments about related topics. I would like to suggest that if bugs about other items have been discovered, new tickets be opened for them instead of recording them here.
